### ISSUE TITLE
Working with dynamic data

### ISSUE THREAD

**Issue by TehWardy** [JamesNK/Newtonsoft.Json#2182]
I'm hitting some issues with some extreme cases in my workflow engine that goes down some deep rabbit holes and hope there's some route for getting around a problem regarding dynamic data.

I have a class that looks something like this ...
```
class WorkflowRequest<T>
{
      public T Data { get; set; }
}
```
... when dealing with this in the workflow engine do a standard deserialize as you would expect ...
```
var request = JsonConvert.DeserializeObject<WorkflowRequest<dynamic>>();
```

... at this point what actually get back is a WorkflowRequest<JObject> instance and that due to the way generics works causes me some issues.

consider this expression ...
```
string x = request.Data.foo;
```
... for a JObject with JValues this doesn't work.

So I need a way to ensure that my initial deserialize actually returned an object in a form that I can consume when we consider that the "expression" is generated code working on what it thinks is a dynamic object but is actually a JObject.

Am I expecting too much here?

In short: is there a way to say to the deserializer "never give me a JObject with JValues, always give me an ExpandoObject or something instead"?

**Comment by TehWardy** [comment]
Probably related to this ...

https://stackoverflow.com/questions/37993788/odata-collection-parameter-bindings-do-they-actually-work/38051214#38051214

... hopefully my experience might help here.
